server/src/main/java/timely/store/iterators/TimeSeriesGroupingIterator.java

```diff
@@ -0,0 +1,309 @@
+import java.util.TreeMap;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.core.iterators.WrappingIterator;
+import org.apache.accumulo.core.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import timely.adapter.accumulo.MetricAdapter;
+import timely.model.Metric;
+
+/**
+ * Iterator that groups time series so that a filter can be applied to the time
+ * series. This iterator will ignore time series that do not have enough data
+ * points at the start of the query.
+ *
+ */
+public class TimeSeriesGroupingIterator extends WrappingIterator {
+
+    /**
+     * Object representing a single time series and its computed value
+     *
+     */
+    private static class TimeSeries extends TreeMap<Key, Double> implements Serializable {
+
+        private static final long serialVersionUID = 1L;
+        private int targetSize;
+        private transient TimeSeriesGroupingIterator iter;
+        private Double answer;
+
+        public TimeSeries(TimeSeriesGroupingIterator iter, int size) {
+            this.iter = iter;
+            this.targetSize = size;
+        }
+
+        @Override
+        public Double put(Key key, Double value) {
+            return this.put(key, value, true);
+        }
+
+        public Double put(Key key, Double value, boolean doCompute) {
+            if (this.targetSize == super.size()) {
+                Entry<Key, Double> e = this.pollFirstEntry();
+                LOG.trace("Removing first entry {}", e.getKey());
+            }
+            Double result = super.put(key, value);
+            if (doCompute) {
+                recompute();
+            }
+            return result;
+        }
+
+        public Double getAnswer() {
+            return this.answer;
+        }
+
+        private void recompute() {
+            this.answer = iter.compute(this);
+            LOG.trace("recomputed, new answer: {}", this.answer);
+        }
+
+    }
+
+    private static class TimeSeriesMetricComparator implements Comparator<Metric>, Serializable {
+
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public int compare(Metric m1, Metric m2) {
+            int result = m1.getName().compareTo(m2.getName());
+            if (result == 0) {
+                result = m1.getTags().toString().compareTo(m2.getTags().toString());
+            }
+            return result;
+        }
+
+    }
+
+    /**
+     *
+     * Object representing a group of time series, where uniqueness is defined
+     * by the name of the metric and a unique tag set.
+     *
+     */
+    private static class TimeSeriesGroup extends TreeMap<Metric, TimeSeries> implements Iterable<Pair<Key, Double>> {
+
```

Comment:
Another minor memory savings is to use a key that only contains the name and the tags.  The Value pointer in the Metric class is wasted memory here.